src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs

```diff
@@ -363,26 +363,25 @@ internal bool TryWrite(Utf8JsonWriter writer, in T value, JsonSerializerOptions
                     // For internal converter only: Handle polymorphic case and get the new converter.
                     // Custom converter, even though polymorphic converter, get called for reading AND writing.
                     JsonConverter jsonConverter = state.Current.InitializeReEntry(type, options);
```

Comment:
In a follow up, we could `Debug.Assert` in `InitializeReEntry` that `type` is not `== typeof(object)`.